util/expensivequery/expensivequery.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expensivequery
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Handle is the handler for expensive query.
+type Handle struct {
+	sctx            sessionctx.Context
+	memExceedConnCh chan uint64
+	exitCh          chan struct{}
+}
+
+// NewExpensiveQueryHandle builds a new expensive query handler.
+func NewExpensiveQueryHandle(sctx sessionctx.Context, exitCh chan struct{}) *Handle {
+	return &Handle{
+		sctx,
+		make(chan uint64),
+		exitCh,
+	}
+}
+
+// Run starts a expensive query checker goroutine at the start time of the server.
+func (eqh *Handle) Run(sm util.SessionManager) {
+	defer close(eqh.memExceedConnCh)
+	sctx := eqh.sctx
+	curInterval := time.Second * time.Duration(sctx.GetSessionVars().ExpensiveQueryTimeThreshold)
+	ticker := time.NewTicker(curInterval / 2)
+	// lastQueryStartTime is used to avoid print duplicated log when exceeding time threshold.
+	lastQueryStartTime := make(map[uint64]time.Time)
+	for {
+		select {
+		case <-ticker.C:
+			if log.GetLevel() > zapcore.WarnLevel {
+				continue
+			}
+			processInfo := sm.ShowProcessList()
+			for connID, info := range processInfo {
+				if len(info.Info) == 0 || info.Time.Equal(lastQueryStartTime[connID]) {
```

Comment:
`lastQueryStartTime` can be removed and `info.Time.Equal(lastQueryStartTime[connID]` can be replaced by checking start time.
If `time.Since(info.Time) >= ExpensiveQueryTimeThreshold + curInterval`, we can assume that we have already logged for this connection and don't need to do it again.

Reply:
Do we need to clear this map periodically?